katrfireport/mkat_rfi_report.py

```diff
@@ -0,0 +1,280 @@
+import logging
+from typing import Dict
+import json
+import os
+
+import katdal
+import numpy as np
+import pandas as pd
+
+import bokeh.embed
+import bokeh.palettes
+import bokeh.plotting
+import bokeh.model
+import bokeh.models
+import bokeh.resources
+from bokeh.io import output_notebook
+from bokeh.layouts import gridplot, column
+from bokeh.models import LinearColorMapper, ColorBar, BasicTicker
+from bokeh.models.widgets import Panel
+from bokeh.models.widgets import Tabs
+from bokeh.models import Range1d
+import warnings
+warnings.filterwarnings('ignore')
+
+
+class PlotFreqTimeStats:
+    """Collect RFI statistics for frequency-time plot."""
+    def __init__(self, dataset: katdal.DataSet, **kwargs) -> None:
+        self.unixtime = dataset.timestamps  # Unix timestamps
+        self.frequency = dataset.freqs/1e6  # Frequency in MHz
+        self.x_range = Range1d(self.frequency.min(), self.frequency.max())
+        self.y_range = Range1d(0, len(self.unixtime))
+
+    @staticmethod
+    def _waterfall_plot(fig: bokeh.plotting.Figure, *args, **kwargs) -> None:
+        fig.image(*args, **kwargs)
+        # Creating a colorbar object
+        color = LinearColorMapper(palette="Viridis256", low=0, high=1)
+        cb = ColorBar(color_mapper=color, location=(5, 6), ticker=BasicTicker())
+        fig.add_layout(cb, 'right')
+
+    def make_rfi_stats_data_source(self, two_d_array) -> bokeh.models.ColumnDataSource:
+        data = {'image': two_d_array}
+        return bokeh.models.ColumnDataSource(data)
+
+    def format_fig(self, title, pol, dataset: katdal.DataSet):
+        fig = bokeh.plotting.figure(
+        x_axis_label='Frequency MHz',
+        y_axis_label=('Pol {} Scans').format(pol),
+        sizing_mode='stretch_width',
+        title=title,
+        width=1000, height=500, toolbar_location='above')
+        return fig
+
+    def freq_time_fig(self, dataset: katdal.DataSet, title, pol,
+                      source: bokeh.models.ColumnDataSource) -> bokeh.model.Model:
+        fig = self.format_fig(title, pol)
+        freqs = self.frequency
+        fig.x_range = self.x_range
+        fig.y_range = self.y_range
+        y_ticks_dic = self.make_ticks(dataset)
+        fig.yaxis.ticker = np.array([*y_ticks_dic], dtype=np.float)
+        fig.yaxis.major_label_overrides = y_ticks_dic
+        self._waterfall_plot(
+            fig, image=[source.data['image']], x=freqs.min(),
+            y=0, dw=freqs.max()-freqs.min(),
+            dh=len(self.unixtime), palette="Viridis256")
+        return fig
+
+    def make_ticks(self, dataset: katdal.DataSet):
+        """Make axis ticks"""
+        # Make yticks dictionary required by bokeh
```

Comment:
I'm concerned that making the ticks like this will cause a mismatch between the tick labels and the actual scan targets, because the scan lengths aren't necessarily consistent. I think a better approach here would be to use something like `dataset.sensor.get('Observation/target')` to build the dictionary because this already lists the timestamps and targets for the observation.